### PR TITLE
Settings Driver

### PR DESCRIPTION
**Pull request by lluwang**
This is one of PRs for 9.2.8.

### CODE REVIEW

File: src/core/common/settings.cpp [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+static uint16_t sStageActionSeq;
+static struct stageAddSettingsBlock sStageAddSettingsBlock[MAX_STAGE_ADD_NUM];
+static uint8_t sStageAddSettingsNum;
+static uint16_t sStageAddSettingsBufLength;
+static struct stageDeleteSettingsBlock sStageDeleteSettingsBlock[MAX_STAGE_DELETE_NUM];
+static uint8_t sStageDeleteSettingsNum;
+
+void reorderSettingsBlock(void)
+{
+    uint8_t *ramBuffer = (uint8_t *)malloc(OPENTHREAD_CONFIG_SETTINGS_PAGE_SIZE);
```
Comment (nibanks) [inline src/core/common/settings.cpp]: You can't call malloc/free directly. When this runs on Windows in kernel mode, you have to call different functions to allocate memory.

File: src/core/common/settings.cpp [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+static uint16_t sStageActionSeq;
+static struct stageAddSettingsBlock sStageAddSettingsBlock[MAX_STAGE_ADD_NUM];
+static uint8_t sStageAddSettingsNum;
+static uint16_t sStageAddSettingsBufLength;
+static struct stageDeleteSettingsBlock sStageDeleteSettingsBlock[MAX_STAGE_DELETE_NUM];
+static uint8_t sStageDeleteSettingsNum;
+
+void reorderSettingsBlock(void)
+{
+    uint8_t ramBuffer[OPENTHREAD_CONFIG_SETTINGS_PAGE_SIZE];
```
Comment (jwhui) [inline src/core/common/settings.cpp]: Buffering settings in RAM to rewrite the settings block does allow this to operate on a single flash page.  However, there are a couple issues with this:

1. If the device resets (or loses the memory) for any reason during this operation, all the settings will be lost.
2. This requires buffering a significant amount of state in memory before writing it back.

I'd prefer a solution that swaps between two different flash pages.  Using two flash pages would avoid needing to buffer all the settings in RAM and can be made resilient to device reboots.

Thoughts?

Comment (luwannest) [inline src/core/common/settings.cpp]: Agree. It is better.

I will use one flash page as the swap page (similar to the RAM space).
The possible problem is that this page may be read/write many more times than others.

Comment (luwannest) [inline src/core/common/settings.cpp]: Yes. But may it need more than one flash page to store the settings?

Comment (luwannest) [inline src/core/common/settings.cpp]: Another way is to always use  2* N flash pages. N 's default is 1, but can be larger like 2, 3....

**Comment by nibanks** [comment]
I've got a more basic question of why OpenThread is dealing with something so platform specific as flash pages to begin with. On Windows this makes no sense. If we need to persist a setting, we'd likely just write it to a registry key. Shouldn't the otPlat* API be something basic like read/write setting? We can move this code for converting settings to flash pages into an intermediate library a platform can optionally use.

**Comment by jwhui** [comment]
The [`otPlatSettings*`](https://github.com/openthread/openthread/blob/master/include/platform/settings.h) API is the only real requirement for OpenThread.  At the same time, we expect a common target for OpenThread are MCUs with on-chip flash storage capability.  For that reason, we've decided to also define a common implementation of `otPlatSettings*` that lives on top of [`otPlatFlash*`](https://github.com/openthread/openthread/blob/master/include/platform/flash.h) APIs.  The intent is that this implementation of `otPlatSettings*` is only provided for convenience and can be replaced on a given platform.  I agree that this PR in its current form  does not completely fulfill the intent.

**Comment by luwannest** [comment]
@jwhui , so would you prefer put settings driver in example/platform/posix and example/platform/cc2538?

**Comment by jwhui** [comment]
@luwannest, we can leave it where it is now.  However, we should provide compile-time options to allow building with an alternative platform-specific implementation of `otPlatSettings*`.  I wouldn't necessarily worry about those compile-time options in this PR, we can introduce those in a separate PR.

**Comment by jwhui** [comment]
@lluwang, thanks for updating this PR.  I'm a little concerned about staging the changes in RAM.  I think we could simplify this by removing support for:

- `otPlatSettingsBeginChange()`
- `otPlatSettingsCommitChange()`
- `otPlatSettingsAbandonChange()`

The above functions are noted as optional in the docs.

Also, we need to update the API to support multiple instances as we have done with the rest of the OpenThread APIs.  At the same time, we need to avoid using globals within the implementation.

File: src/core/common/settings.cpp [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+{
+    kBlockAddFlag = 0x1,
+    kBlockDeleteFlag = 0x02,
+};
+
+enum
+{
+    kMaxStageAddNum = 3,
+    kMaxStageDeleteNum = 3,
+    kMaxStageDataLen = 32,
```
Comment (jwhui) [inline src/core/common/settings.cpp]: Not large enough to support Operational Datasets.

File: src/core/common/settings.cpp [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+static uint32_t sUsedSize;
+static bool sCommitLock = false;
+static bool sSwapDone = false;
+
+static uint16_t sStageActionSeq;
+static struct stageAddSettingsBlock sStageAddSettingsBlock[kMaxStageAddNum];
+static uint8_t sStageAddSettingsNum;
+static uint16_t sStageAddSettingsBufLength;
+static struct stageDeleteSettingsBlock sStageDeleteSettingsBlock[kMaxStageDeleteNum];
+static uint8_t sStageDeleteSettingsNum;
```
Comment (jwhui) [inline src/core/common/settings.cpp]: Avoid using globals.

File: include/platform/settings.h [hunk 5]
```diff
@@ -229,7 +229,7 @@ ThreadError otPlatSettingsAdd(uint16_t aKey, const uint8_t *aValue, int aValueLe
  *  @retval kThreadError_NotImplemented
  *          This function is not implemented on this platform.
  */
-ThreadError otPlatSettingsDelete(uint16_t aKey, int index);
+ThreadError otPlatSettingsDelete(uint16_t aKey, int aIndex);
```
Comment (jwhui) [inline include/platform/settings.h]: Need to update the `otPlatSettings*` APIs to take an `otInstance`.

File: src/core/common/settings.cpp [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+
+    if (!sCommitLock)
+    {
+        if ((sUsedSize + stageBlock->block.length + sizeof(struct settingsBlock)) >= sSettingsSize)
+        {
+            VerifyOrExit(swapSettingsBlock() >= (stageBlock->block.length + sizeof(struct settingsBlock)),
+                         error = kThreadError_NoBufs);
+        }
+
+        otPlatFlashWrite(sBase + sUsedSize, reinterpret_cast<uint8_t *>(stageBlock),
```
Comment (jwhui) [inline src/core/common/settings.cpp]: We need to make sure the write is atomic.  For example, what if the node reboots in the middle of this flash write?  What probably needs to happen here is first write out the block header and data.  Then when that's complete, modify a bit in the header to indicate that the write has completed.

File: src/core/common/settings.cpp [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+    while (address < (sBase + sSettingsSize))
+    {
+        struct settingsBlock block;
+
+        otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+        if ((!(block.flag & kBlockAddFlag) && block.flag & kBlockDeleteFlag) && (block.key == aKey))
+        {
+            SuccessOrExit(error = otPlatSettingsDelete(aKey, -1));
```
Comment (jwhui) [inline src/core/common/settings.cpp]: Should not delete the old setting until the new setting is written.  Maybe a simpler approach here is simply not to delete.  A later block with the same key and index 0 indicates that all prior blocks are invalidated.

File: src/core/common/settings.cpp (hunk 3 above)
Comment (lluwang) [inline src/core/common/settings.cpp]: I have removed otPlatSettingsChangesBegin/otPlatSettingsCommitChanges/otPlatSettingsAbandonChanges.

File: src/core/common/settings.cpp (hunk 4 above)
Comment (lluwang) [inline src/core/common/settings.cpp]: I still keep using sBase and sUsedSize as global variables,and failed to find a way not using them as global...

File: src/core/common/settings.cpp (hunk 7 above)
Comment (lluwang) [inline src/core/common/settings.cpp]: I have changed the logic here.
otPlatSettingsSet is otPlatSettingsAdd with index 0.

The block entry is valid only when
- kSettingsAddComplete flag is not set
- kSettingsDelete flag is set
- there is not block entry containing the same key and <= index after this block entry

File: src/core/common/settings.cpp (hunk 6 above)
Comment (lluwang) [inline src/core/common/settings.cpp]: I split kSettingsAdd flag into kBlockAddBeginFlag and kBlockAddCompleteFlag for atomic write.

File: include/platform/settings.h (hunk 5 above)
Comment (lluwang) [inline include/platform/settings.h]: otInstance has been added for otPlatSettings*.
I will submit a separate PR for otPlatFlash*.

File: src/core/common/settings.cpp [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+{
+    kBlockAddBeginFlag = 0x1,
+    kBlockAddCompleteFlag = 0x02,
+    kBlockDeleteFlag = 0x04,
+};
+
+enum
+{
+    kSettingsFlagSize = 4,
+    kSettingsBlockDataSize = 32,
```
Comment (jwhui) [inline src/core/common/settings.cpp]: You will need to buffer up to 255 bytes to support the Operational Datasets.

File: tests/unit/Makefile.am [hunk 9]
```diff
@@ -128,6 +129,11 @@ test_message_SOURCES         = test_platform.cpp test_message.cpp
 test_ncp_buffer_LDADD        = $(COMMON_LDADD)
 test_ncp_buffer_SOURCES      = test_platform.cpp test_ncp_buffer.cpp
 
+test_settings_LDADD          = $(COMMON_LDADD)  \
+                               $(top_builddir)/examples/platforms/posix/libopenthread-posix.a \
+                               $(NULL)
+test_settings_SOURCES        = test_settings.cpp
```
Comment (nibanks) [inline tests/unit/Makefile.am]: Can we please follow the model of the tests not depending on a specific platform? Can you please include the `test_platform.cpp` as the other tests above do?

Comment (jwhui) [inline tests/unit/Makefile.am]: In this case, the implementation in `settings.cpp` depends on a working implementation of `otPlatFlash*`.  We could include `examples/platforms/posix/flash.c` directly, but that would still mean this test depends on a specific platform.  Thoughts?

Comment (jwhui) [inline tests/unit/Makefile.am]: I think a "mock" in-memory implementation of `otPlatFlash*` is reasonable.  @lluwang, thoughts?

Comment (lluwang) [inline tests/unit/Makefile.am]: But till my understanding, it is almost the same as what posix does except that using in-memory space but not flash?

Comment (jwhui) [inline tests/unit/Makefile.am]: The idea is to remove any platform-specific dependencies (i.e. file operations).

**Comment by lluwang** [comment]
Is there any way for me to get the contents of sFlashBuffer from travis?
I run the test on my local linux machine that all tests are passed, but always failed by travis.

**Comment by lluwang** [comment]
Or is it a bad way to allocate this big (4Kbytes) buffer?

**Comment by jwhui** [comment]
@lluwang, I just tried running `make -f examples/Makefile-posix distcheck` on OSX and I see the same failure as on travis.

**Comment by lluwang** [comment]
Thanks @jwhui !

File: tests/unit/test_settings.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+                 "Settings::Swap::Add Fail after swap\n");
+    VerifyOrQuit(otPlatSettingsGet(&sInstance, key, index - 1, readBuffer, &readBufferLength) == kThreadError_None,
+                 "Settings::Swap::Get Fail\n");
+    VerifyOrQuit(!memcmp(readBuffer, sWriteBuffer, static_cast<uint16_t>(sWriteBufferLength)),
+                 "Settings::Swap::Add and Swap Check Fail\n");
+}
+
+void RunSettingsTests(void)
+{
+    TestSettingsInit();
```
Comment (nibanks) [inline tests/unit/test_settings.cpp]: Can you add these tests to `test_windows.cpp`?

**Comment by nibanks** [comment]
To get appveyor building, you need to fix the warning in `settings.cpp` as shown [here](https://ci.appveyor.com/project/jwhui/openthread/build/0.1.620/job/5gjj00r1cu2sgmb2/messages).

**Comment by nibanks** [comment]
Actually, I think you can just not include `settings.cpp` in libopenthread_k for now, since the kernel mode version of the library is what I use in the driver on Windows, and I will implement my own versions of the otPlat* APIs. But it would be great still to fix those warnings.

**Comment by lluwang** [comment]
Could I know how test_windows.cpp run the unit test?
Will it run TestSettingsInit, TestSettingsAdd, TestSettingsDelete, TestSettingsSet, TestSettingsSwap in sequence?

**Comment by nibanks** [comment]
It looks like it [is running ](https://ci.appveyor.com/project/jwhui/openthread/build/0.1.621/job/djs6j6xdwu2k59x1/tests)your unit tests, and one is failing. Assertion failed at:

`at TestSettingsSwap() in c:\projects\openthread\tests\unit\test_settings.cpp:line 139`

**Comment by nibanks** [comment]
It looks like it doesn't necessarily run the tests in sequence. If the tests have dependencies among each other, they aren't really separate tests then. Either you need to make them independent, or make Windows just run `RunSettingsTests` and not each individual test.

**Comment by jwhui** [comment]
From Seth Rickard \<s-rickard@ti.com\>:

> The area your flash driver is writing begins at 0x00239000 and continues for a few pages. This puts the NV storage area about 228K into the chip’s flash. This is well above the code that is currently in OpenThread, but the area is not reserved in the linker script. It would be possible to write an application using OpenThread that has code in that area, without the linker complaining. The application would end up being accidentally overwritten by the settings.
> 
> The linker script might look like
> ```.ld
> SECTIONS
> {
>     /* place non-volatile storage in pages 510 and 511 */
>     .nv_storage 0x7F800:
>     {
>         _nv_begin = .;
>         . += 0x400;
>         _nv_page_one = .;
>         . += 0x400;
>         _nv_end = .;
>     } > FLASH
> }
> ```
>  
> Then in the c file, finding the addresses would look like
> ```.c
> extern char _nv_begin;
> extern char _nv_page_one;
> extern char _nv_end;
>  
> int someFunc()
> {
>     char *flash_base = &_nv_begin;
>     uint32_t size_of_page = (uint32_t)(&_nv_page_one - &_nv_begin);
>     uint8_t num_of_pages = (uint8_t)((&_nv_end - &_nv_begin) / size_of_page);
> }
> ```
>  
> This way you don’t have to use #define s for the base address and page size

**Comment by lluwang** [comment]
If we do like this, is it specific to some platforms? Though settings driver may always only be used in embedded platform.

For the potential flash error, do we need to have a component to manage the flash layout?

File: etc/visual-studio/libopenthread_k.vcxproj [hunk 11]
```diff
@@ -23,6 +23,7 @@
     <ClCompile Include="..\..\src\core\coap\coap_server.cpp" />
     <ClCompile Include="..\..\src\core\common\logging.cpp" />
     <ClCompile Include="..\..\src\core\common\message.cpp" />
+    <ClCompile Include="..\..\src\core\common\settings.cpp" />
```
Comment (nibanks) [inline etc/visual-studio/libopenthread_k.vcxproj]: Don't add this file to the kernel mode version.

**Comment by jwhui** [comment]
LGTM 👍 , we can address Seth's concerns in a separate PR.

**Pull request merged**